kernel/src/sched.rs

```diff
@@ -331,6 +336,7 @@ impl Kernel {
         systick.reset();
         systick.set_timer(KERNEL_TICK_DURATION_US);
         systick.enable(false);
+        let identifier: usize = process.appid().id();
```

Comment:
```suggestion
        let identifier = process.appid();
```

Ah, I see where this is a bit non-intuitive, but the plain `appid` should be used in-kernel; the `.id()` method is only for special situations: https://docs.tockos.org/kernel/struct.appid#method.id

Reply:
This is subtle, but we should never make a copy of a process's ID like this, since it can change (see 4003b297c90f693ea297a3c70de35f28a0d49c7f for an example of where this has lead to a bug in the past). Instead just use `process.appid()` when needed.

Reply:
Thank you very much for reviewing this @ppannuto and @bradjc! Your comments raise two very important points:

**Caching process IDs should, in general, be avoided**
Thanks for pointing me to #1613 @bradjc! It is indeed a subtle bug caused by stale process IDs. However, this PR actually relies on the fact that the process IDs returned by `process.appid().id()` is unique among different processes and different _instances_ of the _same_ process [1]. This ensures that the MPU will be configured whenever we switch to a different process. I guess, another alternative could be to store this information as a flag in the process structure itself, but I find this approach more intuitive.

**`process.appid()` vs `process.appid().id()`**
@ppannuto, the reason for using `process.appid().id()` was performance. Using `process.appid()` introduces an overhead of 150 cycles.

[1]  Apologies for the abuse of terminology. Here, I am referring to processes that are restarted after crashes or updated at runtime.

Reply:
I'm afraid this does the opposite of what you might want. It is perfectly valid for the `.id()` value to be re-used. In fact, right now it is simply the index into the static array of `Process` pointers. If a process restarts, `.appid()` will change and `.appid().id()` will not.

However, if a process is restarted, its memory layout likely _has_ changed (i.e. no `brk` yet), thus the MPU regions should be reconfigured. I believe `.appid()` is still the correct choice, as it is what is unique among instances of a process.

Reply:
Oh, I was under the impression that `index` can be reused but `.id()`, i.e. `identifier`, is unique because `create_process_identifier()` is called when a process is created or restarted.

Reply:
Yes, sorry, you are right. I described the old behavior (id used to be the index; #1566 updated this).

Nonetheless, comparing AppId should do the right thing (it internally compares the identifiers returned by `.id()`):
```rust
impl PartialEq for AppId {
    fn eq(&self, other: &AppId) -> bool {
        self.identifier == other.identifier
    }
}
```

---

I know the _intent_ of the `AppId`/`identifier` split was to use `AppId`s internally and to provide `identifier`s for external use, but I'm not sure I could articulate when/where/why i'ts preferred to use `AppId` over the `identifer`

Reply:
No worries, sounds good! I will update this PR to use `AppId` (we can hopefully avoid the overhead by using `MapCell` instead of `Cell`).